File: src/renderer/src/services/connect/lastfm.ts
import { fetchConnect } from './http'

/** One track as Last.fm names it — plain text, not yet anything this
 * library owns. Turning these into songs is lastfmMatcher.ts's job. */
export interface LastfmTrack {
  title: string
  artist: string
  /** Last.fm's recording MBID where it has one. Often absent, and not
   * reliable enough to identify a track by on its own - see
   * core/lastfm.py's _to_tracks(). */
  mbid: string
}

export type LastfmOp = 'charts' | 'genre' | 'artist' | 'similar' | 'mytop'

/** The periods user.getTopTracks accepts, validated by the backend too
 * (routes/lastfm.py's Period). */
export type LastfmPeriod = 'overall' | '7day' | '1month' | '3month' | '6month' | '12month'

export interface LastfmQuery {
  op: LastfmOp
  limit: number
  /** charts only - empty means the global chart rather than a country's. */
  country?: string
  /** genre only. */
  tag?: string
  /** artist only. */
  artist?: string
  /** similar only - the seed track's title, together with `artist`. */
  track?: string
  /** mytop only - a public Last.fm username. No login: user.getTopTracks
   * takes the name as a plain parameter. */
  username?: string
  /** mytop only. */
  period?: LastfmPeriod
}

export async function getLastfmTracks(query: LastfmQuery): Promise<LastfmTrack[]> {
  const params = new URLSearchParams({ op: query.op, limit: String(query.limit) })
  if (query.country) params.set('country', query.country)
  if (query.tag) params.set('tag', query.tag)
  if (query.artist) params.set('artist', query.artist)
  if (query.track) params.set('track', query.track)
  if (query.username) params.set('username', query.username)
  if (query.period) params.set('period', query.period)
  const data = await fetchConnect<{ tracks: LastfmTrack[] }>(`/lastfm/songs?${params.toString()}`)
  return data.tracks
}

File: connect/routes/lastfm.py
"""routes/lastfm.py — GET /lastfm/songs

Machine-to-machine (CONNECT_TOKEN), not session-scoped: like
routes/recommendations.py, nothing here touches session.media. The answer
is a list of title/artist names straight from Last.fm — finding the ones
that exist in the library is the renderer's job (see core/lastfm.py's
docstring for why).

The key this needs is read from core/api_keys.py; managing it (status,
storing, clearing) lives in routes/api_keys.py for every keyed service at
once, not here.
"""

import logging
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, Query

from core import lastfm
from core.auth import require_token

logger = logging.getLogger("connect.lastfm")
router = APIRouter(prefix="/lastfm", dependencies=[Depends(require_token)])

# The period values user.getTopTracks accepts. Validated here rather than
# passed through so a typo comes back as a 422 naming the alternatives,
# instead of Last.fm's generic "invalid parameters" error 6.
Period = Literal["overall", "7day", "1month", "3month", "6month", "12month"]


# The path says "songs" while everything else here says "tracks": ad-blocker
# filter lists flag a /tracks segment as a tracker endpoint, which had the
# web build's requests blocked before they reached connect. The payload is a
# track list either way.
@router.get("/songs")
async def tracks(
    op: Literal["charts", "genre", "artist", "mytop"],
    limit: int = Query(default=50, ge=1, le=500),
    country: str = "",
    tag: str = "",
    artist: str = "",
    username: str = "",
    period: Period = "1month",
) -> dict:
    if not lastfm.is_configured():
        raise HTTPException(
            status_code=503,
            detail="No Last.fm API key configured for this installation",
        )

    try:
        if op == "charts":
            # No country means the global chart; the frontend sends one or
            # the other rather than a separate op for each.
            if country:
                result = await lastfm.get_geo_top_tracks(country, limit)
            else:
                result = await lastfm.get_global_top_tracks(limit)
        elif op == "genre":
            if not tag:
                raise HTTPException(status_code=422, detail="genre needs a tag")
            result = await lastfm.get_tag_top_tracks(tag, limit)
        elif op == "artist":
            if not artist:
                raise HTTPException(status_code=422, detail="artist needs an artist name")
            result = await lastfm.get_artist_top_tracks(artist, limit)
        else:
            if not username:
                raise HTTPException(status_code=422, detail="mytop needs a Last.fm username")
            result = await lastfm.get_user_top_tracks(username, period, limit)
    except lastfm.LastfmError as e:
        # 404 for something the user can correct (an unknown username,
        # artist or tag), 502 for anything upstream they cannot.
        raise HTTPException(status_code=404 if e.not_found else 502, detail=str(e)) from e

    return {"tracks": result}
